cmd/fil-daggregator/main.go

```diff
@@ -0,0 +1,294 @@
+		maxWorkers--
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			api := ipfsapi.NewShell(opts.IpfsAPI)
+			api.SetTimeout(time.Second * time.Duration(opts.IpfsAPITimeoutSecs))
+
+			for {
+				toAggIdx, chanOpen := <-workCh
+				if !chanOpen {
+					return
+				}
+
+				ds := new(dagStat)
+				err := api.Request("dag/stat").Arguments(toAgg[toAggIdx].RootCid.String()).Option("progress", "false").Exec(ctx, ds)
+				if err != nil {
+					errCh <- err
+					return
+				}
+
+				toAgg[toAggIdx].UniqueBlockCount = ds.NumBlocks
+				toAgg[toAggIdx].UniqueBlockCumulativeSize = ds.Size
+			}
+		}()
+	}
+
+	wg.Wait()
+	shutdownWorkers()
+
+	errCh <- externalCtx.Err()
+	return <-errCh
+}
+
+// pulls cids from an AllKeysChan and sends them concurrently via multiple workers to an API
+func writeoutBlocks(externalCtx context.Context, opts *opts, bs blockstore.Blockstore) error {
+
+	ctx, shutdownWorkers := context.WithCancel(externalCtx)
+	defer shutdownWorkers()
+
+	akc, err := bs.AllKeysChan(ctx)
+	if err != nil {
+		return err
+	}
+
+	maxWorkers := opts.IpfsAPIMaxWorkers
+	errCh := make(chan error, 1+maxWorkers) // our own error plus one from each worker
+
+	// watchdog to bail early if an error appears
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				errCh <- ctx.Err()
+				return
+			case err := <-errCh:
+				errCh <- err // put error back where we got it from
+				if err != nil {
+					shutdownWorkers()
+				}
+				return
+			}
+		}
+	}()
+
+	var wg sync.WaitGroup
+
+	for maxWorkers > 0 {
+		maxWorkers--
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			api := ipfsapi.NewShell(opts.IpfsAPI)
+			api.SetTimeout(time.Second * time.Duration(opts.IpfsAPITimeoutSecs))
+
+			for {
+				select {
+
+				case <-ctx.Done():
+					return
+
+				case c, chanOpen := <-akc:
+
+					if !chanOpen {
+						return
+					}
+
+					blk, err := bs.Get(c)
+					if err != nil {
+						errCh <- err
+						return
+					}
+
+					// copied entirety of ipfsapi.BlockPut() to be able to pass in our own ctx 🤮
+					res := new(struct{ Key string })
+					err = api.Request("block/put").
+						Option("format", cid.CodecToStr[c.Prefix().Codec]).
+						Option("mhtype", multihash.Codes[c.Prefix().MhType]).
+						Option("mhlen", c.Prefix().MhLength).
+						Body(
+							ipfsfiles.NewMultiFileReader(
+								ipfsfiles.NewSliceDirectory([]ipfsfiles.DirEntry{
+									ipfsfiles.FileEntry(
+										"",
+										ipfsfiles.NewBytesFile(blk.RawData()),
+									),
+								}),
+								true,
+							),
+						).
+						Exec(ctx, res)
+					// end of 🤮
+
+					if err != nil {
+						errCh <- err
+						return
+					}
+
+					if res.Key != c.String() {
+						errCh <- xerrors.Errorf("unexpected cid mismatch after /block/put: expected %s but got %s", c, res.Key)
+						return
+					}
+				}
+			}
+		}()
+	}
+
+	wg.Wait()
+	shutdownWorkers()
+
+	// feeder drain if anything remains
+	for len(akc) > 0 {
+		<-akc
+	}
+
+	errCh <- externalCtx.Err()
+	return <-errCh
+}
```

Comment:
@mvdan since your head is in this space these days - could you please look over this entire function? It seems **way too** fiddly to me, but at the same time I don't see how I can make it simpler... ( I do not have much experience with heavy-async/nested-context work )